pkg/services/ngalert/provisioning/provenance.go

```diff
@@ -11,3 +11,18 @@ func canUpdateProvenanceInRuleGroup(storedProvenance, provenance models.Provenan
 		storedProvenance == models.ProvenanceNone ||
 		(storedProvenance == models.ProvenanceAPI && provenance == models.ProvenanceNone)
 }
+
+type ProvenanceStatusTransitionValidator = func(from, to models.Provenance) error
+
+// ValidateProvenanceRelaxed checks if the transition of provenance status from `from` to `to` is allowed.
+// Applies relaxed checks that prevents only transition from any status to `none`.
+// Returns ErrProvenanceChangeNotAllowed if transition is not allowed
+func ValidateProvenanceRelaxed(from, to models.Provenance) error {
```

Comment:
I like this code design overall, but the logic here is missing support for `X-Disable-Provenance`. We allow going from `ProvenanceAPI` to `ProvenanceNone` iff that header is set.

The way the code surrounding provenance is designed obscures this fact as we're kind of overloading `ProvenanceNone` to both mean "no provenance" and also "disable provenance if we're coming from provisioning API". Feel free to improve it in this PR or leave it to a separate one as you prefer, since it's a bit out of scope here.

Reply:
that header is terminated in the request handler https://github.com/grafana/grafana/blob/5b1498f98f3e1e386381d4e8e630cea75653c5a9/pkg/services/ngalert/api/api_provisioning.go#L295 and then that model is passed to the domain service, which then checks whether provenance is allowed. This is how it's done in other places like contact point https://github.com/grafana/grafana/blob/5b1498f98f3e1e386381d4e8e630cea75653c5a9/pkg/services/ngalert/provisioning/contactpoints.go#L250

I do agree that we can improve the way how the header is treated. However, I think we need to unify the behavior across all services first, and then change the logic in one place.